jlm/util/GraphWriter.hpp

```diff
@@ -0,0 +1,795 @@
+#include <vector>
+
+namespace jlm::util
+{
+
+enum class GraphOutputFormat
+{
+  Dot,   // prints
+  ASCII, // output format that makes edges implicit when possible
+};
+
+enum class AttributeOutputFormat
+{
+  SpaceSeparatedList, // printed on the form attr=value other="value 2"
+  HTMLAttributes      // adds extra restrictions on attribute names
+};
+
+class GraphWriter;
+class Node;
+class Edge;
+class Port;
+class Graph;
+
+class GraphElement
+{
+public:
+  virtual ~GraphElement() = default;
+
+  /**
+   * Constructs a graph element with no label, attributes or associated program object
+   * @param label
+   */
+  GraphElement();
+
+  GraphElement(const GraphElement & other) = delete;
+
+  GraphElement(GraphElement && other) = delete;
+
+  GraphElement &
+  operator=(const GraphElement & other) = delete;
+
+  GraphElement &
+  operator=(GraphElement && other) = delete;
+
+  /**
+   * Gets a short string that will serve as the base for a unique ID.
+   * This base should be a valid C-like identifier.
+   * @return a string, such as "node", "i", "o", "graph"
+   */
+  [[nodiscard]] virtual const char *
+  GetIdStub() const = 0;
+
+  /**
+   * Gives the final unique ID of the GraphElement, such as "node3".
+   * Requires the GraphElement to be finalized.
+   * @return the full id of the GraphElement, including unique suffix
+   */
+  [[nodiscard]] std::string
+  GetFullId() const;
+
+  /**
+   * Gets a reference to the graph this GraphElement belongs to
+   */
+  [[nodiscard]] virtual Graph &
+  GetGraph() = 0;
+
+  /**
+   * Sets the element's label.
+   * A label is text intended to be visible in all renderings of the graph.
+   * Use an empty string to signify no label.
+   * @param label the new label string
+   */
+  void
+  SetLabel(std::string label);
+
+  /**
+   * @return true if this graph element has a non-empty label
+   */
+  [[nodiscard]] bool
+  HasLabel() const;
+
+  /**
+   * @return the GraphElement's label
+   */
+  [[nodiscard]] const std::string &
+  GetLabel() const;
+
+  /**
+   * @return the graph element's label, or if it is empty, the string \p otherwise
+   */
+  [[nodiscard]] const char *
+  GetLabelOr(const char * otherwise) const;
+
+  /**
+   * @return the unique suffix assigned to this element when finalized.
+   */
+  [[nodiscard]] int
+  GetUniqueIdSuffix() const;
+
+  /**
+   * Graph elements often represent objects from the program.
+   * By making this association explicit, GraphElements can be looked up by program object.
+   * When referencing a program object that is associated with a GraphElement,
+   * the unique id of the element is used instead of the address of the program object.
+   * Within a graph, only one graph element can be associated with any given program object.
+   * @param object the object to associate this GraphElement with
+   */
+  template<typename T>
+  void
+  SetProgramObject(T * object)
+  {
+    SetProgramObjectUintptr(reinterpret_cast<uintptr_t>(object));
+  }
+
```

Comment:
Would it be possible to just use void pointer here instead of the template parameter?